### DISCUSSION TITLE
How to adapt jsonforms in a custom UIExtension?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-emfcloud/emfcloud#147] [All other] [answered]
Hi,

I am working on a custom GLSP `AbstractUIExtension` to display Data from a selected element in a property panel placed within the diagram panel:

![image](https://user-images.githubusercontent.com/1339124/158617255-d3cad57a-d517-4e9a-95a5-12e7eed0e70d.png)

The goal is to be independent from Theia for the moment and use the same behaviour in VS-Code or Eclipse. This is why I can not use the Theia tree-editor or jsonPropertyPanelview.

When a user selects a element I want to show a form with all element attributes similar to the theia-tree-editor or the JSON-property-view within the body of my UIExtension (the red area in the picture is part of the diagram pane and not a separate Theia view)

I have already implemented a `GLSPSelectionListener` to update the form depending on the selected element. The code looks like this:

```javascript
		if (this.selectionService.isSingleSelection()) {
			const element=root.index.getById(selectedElements[0]);
			if (element instanceof TaskNode) {
				const task: TaskNode=element;
				// render Task Form
				......
				this.bodyDiv?.appendChild(myNewJSONForm);
			}
			if (element instanceof GatewayNode) {
				// render Gateway Form
				......
			}
			if (element instanceof EventNode) {
				// render Event Form
				......     
			}
		}
```

Can someone help me with a code example how I can adapt the jsonforms in this situation? The tree-editor and the jsonpropertyview is so hard bundled to theia specifica, that I can not extract the core requirement:

 - render a jsonform (I understand this requires React)
 - and link the form with my selected element, so that updates are stored directly to the element (or a GLSP Action is generated to send the data to the server - if necessary)

I wonder how this is possible in an easy way?

I am thankful for any help

===
Ralph

**Comment by ndoschek** [comment, marked as answer]
Hi @rsoika!

The necessary dependencies you will have to add will be:

```
"dependencies": {
    "@jsonforms/core": "^3.0.0-beta.1",
    "@jsonforms/react": "^3.0.0-beta.1",
    "@jsonforms/vanilla-renderers": "^3.0.0-beta.1",
   ...
},
"peerDependencies": {
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    ...
}
```

To render the JSON forms component you need a container node (some `HTMLElement`, which could probably be the `containerElement` in your UIExtension or a child of it) to render it in, as it is done for example in the [`JsonFormsPropertyViewWidget#renderForms`](https://github.com/eclipse-emfcloud/jsonforms-property-view/blob/30ecf1153cd1d77d6699fdcc8e9e4f50f77c4d32/jsonforms-property-view/src/browser/widget.tsx#L102) function.
Also, it will be necessary to change your extension's [file extension](https://www.typescriptlang.org/docs/handbook/jsx.html) to `.tsx` and set `"jsx": "react"` in your `tsconfig.json` to be able to use React (probably you will also have to update your typescript version to >= 4, e.g. `"typescript": "~4.5.5"`).

You hand over the `data`, i.e. your element information as JSONObject.
For the renderer parts you can use the vanilla implementations provided by JSONForms.
In the first step, you can also skip schemas, style context and change handler, then it would look something like this:

```
ReactDOM.render(
    <JsonForms
        data={elementDataAsJson}
        cells={vanillaCells}
        renderers={vanillaRenderers}
        readonly
    />,
    this.containerElement
);
```

I tested it quickly and this snippet renders a simple readonly JSON Forms component in an AbstractUIExtension for a GLSP diagram.
Hope that works out for you as well!

Best wishes,
Nina

PS: I like the idea of integrating it as UIExtension, it looks really cool already!

**Reply by rsoika** [reply]
Thanks a lot for your answer and code snippets. I will go ahead with this and I'll give you my feedback as soon as I succeed or get stuck again ;-)

**Reply by rsoika** [reply]
It looks like I am now running into some problems with the dependencies. As soon as I try to add a dummy method using the `ReactDOM.render` method:

```javascript
import * as React from 'react';
import * as ReactDOM from 'react-dom';
export function renderEmptyForms(node: HTMLElement): void {
	ReactDOM.render(
		<React.Fragment>Please select an element </React.Fragment>,
            node
	);
}
```

I got errors during build which I can not resolve:

```
@open-bpmn/open-bpmn-glsp: yarn run v1.22.17
@open-bpmn/open-bpmn-glsp: $ yarn clean && yarn build && yarn lint
@open-bpmn/open-bpmn-glsp: $ rimraf lib
@open-bpmn/open-bpmn-glsp: $ tsc
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(108,12): error TS2345: Argument of type 'Element' is not assignable to parameter of type 'VNode'.
@open-bpmn/open-bpmn-glsp:   Type 'ReactElement<any, any>' is missing the following properties from type 'VNode': sel, data, children, elm, text
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(110,3): error TS2322: Type 'Element' is not assignable to type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(168,12): error TS2345: Argument of type 'Element' is not assignable to parameter of type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(170,3): error TS2322: Type 'Element' is not assignable to type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(224,12): error TS2345: Argument of type 'Element' is not assignable to parameter of type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-element-views.tsx(226,3): error TS2322: Type 'Element' is not assignable to type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-routing-views.tsx(57,20): error TS2345: Argument of type 'Element' is not assignable to parameter of type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/bpmn-routing-views.tsx(119,3): error TS2322: Type 'Element' is not assignable to type 'VNode'.
@open-bpmn/open-bpmn-glsp: src/property-panel/form-section.tsx(46,4): error TS2708: Cannot use namespace 'JSX' as a value.
@open-bpmn/open-bpmn-glsp: error Command failed with exit code 2.
@open-bpmn/open-bpmn-glsp: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@open-bpmn/open-bpmn-glsp: error Command failed with exit code 2.
@open-bpmn/open-bpmn-glsp: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run prepare exited 2 in '@open-bpmn/open-bpmn-glsp'
```

The strange thing is, that the error message points to a different source code file with a code using svg and JSX - like this:

```javascript
		const vnode = (
			<g>
				<path
					transform={'scale(1.15),translate(0.75,0.75)'}
					// eslint-disable-next-line max-len
					d={icon}
				/>
				<rect class-icon-background x={0} y={0} width={25} height={20} />
				{context.renderChildren(element)}
			</g>
		);
		const subType = getSubType(element);
		if (subType) {
			setAttr(vnode, 'class', subType);
		}
		return vnode;
```

Maybe I have generated now a conflict in my tsconfig.json which looks like this:

```
{
  "extends": "@eclipse-glsp/ts-config/tsconfig.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "baseUrl": ".",
    "reactNamespace": "JSX",
    "jsx": "react",
    "skipLibCheck": true,
    "types": []
  },
  "include": ["src"]
}
```

I have added the following lines some time ago to solve a similar compile issue with the SVG parts of my code:

```
    "reactNamespace": "JSX",
    "types": []
```

Is it possible that I have now generated a conflict with the option `"jsx": "react"`, or do you think it's something different?
I have not really a good understanding of these compilerOptions... :-(

**Reply by ndoschek** [reply]
Hi @rsoika,

In some projects we had the same issue, hence we used a workaround and typed the elements explicitly with `any` to avoid compiling problems.
Please also see for example: [sprotty issue #178](https://github.com/eclipse/sprotty/issues/178) (no described solution did not work in our case).

```
const vnode : any = (
   <g>
      ...
   </g>
);
```

HTH and best wishes,
Nina

**Reply by rsoika** [reply]
yes, this trick solved the compilation errors. Thanks!

**Comment by rsoika** [comment]
@ndoschek I still don't get the React and JsonForms working. I feel like I am in a dark room searching the light switch. I hope you can help me out.

I managed to organize my dependencies, I switched from .ts to .tsx and I am able to compile without errors. But the react part is not working. 

The following code fragment shows 3 different variants I try to render my property panel on a select action (Task, Gateway, Event)

```javascript
	/*
	 * This mehtod reacts on the selection of a BPMN element
	 * and updates the property panel input fields
	 */
	selectionChanged(root: Readonly<SModelRoot>, selectedElements: string[]): void {
		console.log('selection change  received:', root, selectedElements);
		if (this.selectionService.isSingleSelection()) {
			const element = root.index.getById(selectedElements[0]);

			if (element instanceof TaskNode) {
				console.log('...Task selected');
				const task: TaskNode = element;
				console.log('...category=' + task.category + ' documentation ' + task.documentation);
				const documentationInput = document.createElement('input');
				documentationInput.insertAdjacentText('beforeend', '' + task.documentation);
				documentationInput.onchange = function (this) {
					console.log('...do something... ');
				};
				this.bodyDiv?.appendChild(documentationInput);
			}

			if (element instanceof GatewayNode) {
				console.log('...Gateway selected');
				const gateway: GatewayNode = element;
				const elementDataAsJson = {};
				ReactDOM.render(
					<JsonForms
						data={elementDataAsJson}
						cells={vanillaCells}
						renderers={vanillaRenderers}
						readonly
					/>,
					this.bodyDiv
				);
				console.log('...eventtype=' + gateway.category);
			}

			if (element instanceof EventNode) {
				console.log('...Event selected....');
				const event: EventNode = element;
				ReactDOM.render(
					<React.Fragment>Please select an element </React.Fragment>,
					this.bodyDiv
				);
				console.log('...eventtype=' + event.category);
			}
		}
	}
```

The first one is working of course as this code is not using React so far. 

The second one throws the following exception:

```
root INFO ...Gateway selected
root ERROR 5:23:36 PM bpmn-diagram_0 GLSPCommandStack: Failed to execute command: Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
    at Object.throwInvalidHookError (http://localhost:3000/bundle.js:105493:13)
    at useMemo (http://localhost:3000/bundle.js:131484:21)
    at JsonForms (http://localhost:3000/vendors-node_modules_eclipse-glsp_theia-integration_lib_browser_index_js-node_modules_jsonfor-3d7753.bundle.js:10702:71)
    at Module.jsx (http://localhost:3000/vendors-node_modules_sprotty-theia_lib_theia_diagram-module_js.bundle.js:1210:16)
    at Object.createElement (http://localhost:3000/vendors-node_modules_sprotty-theia_lib_theia_diagram-module_js.bundle.js:20174:27)
    at BPMNPropertyPanel.selectionChanged (http://localhost:3000/open-bpmn-theia_lib_browser_frontend-module_js-data_image_svg_xml_charset_utf-8_3Csvg_20xmlns-fdcebc.bundle.js:1069:37)
    at http://localhost:3000/vendors-node_modules_eclipse-glsp_theia-integration_lib_browser_diagram_glsp-diagram-manager_-a838e9.bundle.js:5731:62
    at Array.forEach (<anonymous>)
    at SelectionService.notifyListeners (http://localhost:3000/vendors-node_modules_eclipse-glsp_theia-integration_lib_browser_diagram_glsp-diagram-manager_-a838e9.bundle.js:5731:33)
    at SelectionService.updateSelection (http://localhost:3000/vendors-node_modules_eclipse-glsp_theia-integration_lib_browser_diagram_glsp-diagram-manager_-a838e9.bundle.js:5724:18)
```

I have no idea what this message could mean. 

The third variant, which is only using React to isolate the problem throws a different exception:

```
root INFO ...Event selected....
root ERROR 5:25:03 PM bpmn-diagram_0 GLSPCommandStack: Failed to execute command: Error: Objects are not valid as a React child (found: object with keys {sel, data, children, text, elm, key}). If you meant to render a collection of children, use an array instead.
    at throwOnInvalidObjectType (http://localhost:3000/bundle.js:104182:15)
    at reconcileChildFibers (http://localhost:3000/bundle.js:105082:7)
    at reconcileChildren (http://localhost:3000/bundle.js:107538:28)
    at updateHostRoot (http://localhost:3000/bundle.js:108027:5)
    at beginWork (http://localhost:3000/bundle.js:109393:14)
    at HTMLUnknownElement.callCallback (http://localhost:3000/bundle.js:90957:14)
    at Object.invokeGuardedCallbackDev (http://localhost:3000/bundle.js:91006:16)
    at invokeGuardedCallback (http://localhost:3000/bundle.js:91061:31)
    at beginWork$1 (http://localhost:3000/bundle.js:113972:7)
    at performUnitOfWork (http://localhost:3000/bundle.js:112926:12)
```

I read a lot about this kind of exception, but was unable to solve it. 

I can imaging that my import instructions can be a root of my problems. 

```javascript
import {
	AbstractUIExtension,
	EnableDefaultToolsAction,
	EnableToolsAction,
	SetUIExtensionVisibilityAction,
	ActionDispatcher,
	SModelRoot,
	svg,
	TYPES
} from 'sprotty';
import { codiconCSSClasses } from 'sprotty/lib/utils/codicon';
import {
	Action,
	EditModeListener,
	EditorContextService
} from '@eclipse-glsp/client';

import {
	SelectionListener,
	SelectionService
} from '@eclipse-glsp/client/lib/features/select/selection-service';

import { TaskNode, GatewayNode, EventNode } from '../model';

// Import Instruction sReact and JsonForms
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { JsonForms } from '@jsonforms/react';
import {
	vanillaCells,
	vanillaRenderers
} from '@jsonforms/vanilla-renderers';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const JSX = { createElement: svg };
```


Especiallly the last line my be a problem?

```javascript
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const JSX = { createElement: svg };
```
I think it was a advice from @planger  some time ago to solve a problem using svg in my graphical elements. 
In my UIExtension I am not using svg but if I do not add this magic line of code I got the error:

    Cannot use namespace 'JSX' as a value

And this may be an issue of my compiler settings in the `tsconfig.json`: 

```
{
  "extends": "@eclipse-glsp/ts-config/tsconfig.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "baseUrl": ".",
    "reactNamespace": "JSX",
    "skipLibCheck": true,
    "types": []
  },
  "include": ["src"]
```

I am sorry for asking such kind of stupid questions, but I can't figure out how to solve this by myself. I looks like it was a big fault to ignore javascript the last 15 years... ;-)

**Reply by ndoschek** [reply]
Hi @rsoika,

my assumption is that you are using multiple React versions, which can cause your posted exceptions and also the follow-up issues.
This would be shown by multiple blocks for the same dependency in the `yarn.lock` file.
To overcome this, you can resolve the versions in the client root `package.json` like so:
```
 "resolutions": {
    "**/react": "^16.12.0",
    "**/react-dom": "^16.12.0",
    ...
  }
```

Regarding your imports, they look good to me - we use the same structure and it should work that way if you use `svg` in your file (mostly that will be the case for GLSP element views).

HTH and best wishes,
Nina

**Reply by rsoika** [reply]
Thanks for you feedback. I now started to reorganize my project and split it up into more separate modules with the hope to get rid of to many dependencies (react + sprotty) in the same module - which is currently the case. 
As a result I hope I can separate my property view from the GLSP element views to avoid conflicts with sprotty when I use the jsonForms 

I created a dummy-guideline (for myself): https://github.com/imixs/open-bpmn/blob/master/doc/MODULE_SYSTEM.md
...and I started to read another book about node.js which helps me a lot : https://www.packtpub.com/product/node-js-design-patterns-third-edition/9781839214110

**Comment by rsoika** [comment]
@ndoschek Hooray, I finally made it! Thanks to your help!!

Isolating all the property-panel stuff into a separate module solved all the compiling errors and also all runtime errors. Now I can fill my property panel with `ReactDOM.render` during selection of an element.

![image](https://user-images.githubusercontent.com/1339124/159328131-b61c4818-bf9d-46d5-bd00-4cf786757015.png)

And also `ReactDOM.render(<JsonForms.... )` no longer generates runtime exceptions. It did not yet show any content but this may be another problem.

I now have one more request for help: Can you please take a quick look at the following module code:

https://github.com/imixs/open-bpmn/tree/issue-18/open-bpmn.glsp-client/open-bpmn-properties

Especially on the way I export my `bpmnPropertyModule` from the index.ts file:

https://github.com/imixs/open-bpmn/blob/a87200ab18fdd1b16d1d2156f1d75772a523d9be/open-bpmn.glsp-client/open-bpmn-properties/src/index.ts#L20

Is this the correct way I should design UIExtensions in GLSP? I saw that you mostly use the di.config.ts file instead of a index.ts file? But in my module package.json file I refer to lib/index

```
...
  "main": "lib/index",
  "types": "lib/index"
...
```

In my main GLSP module I now simply import the module and add it as a container module:

```javascript
import {bpmnPropertyModule} from '@open-bpmn/open-bpmn-properties';

const bpmnDiagramModule = new ContainerModule((bind, unbind, isBound, rebind) => {
.....
});

export default function createContainer(widgetId: string): Container {
    // add the BPMN propertyViewModule...
    const container = createClientContainer(bpmnDiagramModule, directTaskEditor,bpmnPropertyModule);
    overrideViewerOptions(container, {
        baseDiv: widgetId,
        hiddenDiv: widgetId + '_hidden'
    });
    return container;
}
```
Would you agree to this approach?

**Comment by tortmayr** [comment]
Hi @rsoika 
Great news! Happy to hear that you finally managed to get everything working. Keeping react dependencies clearly separated from the core glsp/sprotty code is definitely the  way to go.
>Is this the correct way I should design UIExtensions in GLSP? I saw that you mostly use the di.config.ts file instead of a index.ts file? But in my module package.json file I refer to lib/index

Sure that approach is perfectly fine. If you only have only single DI module in your package, exporting it via the main `index.ts` definitely makes sense.
In GLSP/Sprotty we have multiple DI modules in one package. To avoid circular dependencies we typically export each module in its own separate configuration file. You could name that configuration file however you like. We simply settled for the "di.config.ts"  convention to make it easier to identify which files export DI modules.
